p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
-func newPeerStore() *peerStore {
-	return &peerStore{
-		peers:   map[NodeID]*peerInfo{},
-		claimed: map[NodeID]bool{},
+// Subscribe subscribes to peer updates. The caller must consume the peer
+// updates in a timely fashion and close the subscription when done, since
+// delivery is guaranteed and will block peer connection/disconnection
+// otherwise.
+func (m *peerManager) Subscribe() *PeerUpdatesCh {
+	// FIXME: We may want to use a size 1 buffer here. When the router
+	// broadcasts a peer update it has to loop over all of the
+	// subscriptions, and we want to avoid blocking and waiting for a
+	// context switch before continuing to the next subscription. This also
+	// prevents tail latencies from compounding across updates. We also want
+	// to make sure the subscribers are reasonably in sync, so it should be
+	// kept at 1. However, this should be benchmarked first.
+	peerUpdates := NewPeerUpdates(make(chan PeerUpdate))
+	m.mtx.Lock()
+	m.subscriptions[peerUpdates] = peerUpdates
+	m.mtx.Unlock()
+
+	go func() {
+		<-peerUpdates.Done()
+		m.mtx.Lock()
+		delete(m.subscriptions, peerUpdates)
+		m.mtx.Unlock()
+	}()
+	return peerUpdates
+}
+
+// broadcast broadcasts a peer update to all subscriptions. The caller must
+// already hold the mutex lock. This means the mutex is held for the duration
+// of the broadcast, which we want to make sure all subscriptions receive all
+// updates in the same order.
+//
+// FIXME: Consider using more fine-grained mutexes here, and/or a channel to
+// enforce ordering of updates.
+func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
+	for _, sub := range m.subscriptions {
+		select {
+		case sub.updatesCh <- peerUpdate:
+		case <-sub.doneCh:
+		}
 	}
 }
 
-// Add adds a peer to the store, given as an address.
-func (s *peerStore) Add(address PeerAddress) error {
-	if err := address.Validate(); err != nil {
+// DialNext finds an appropriate peer address to dial, and marks it as dialing.
+// The peer will not be returned again until Dialed() or DialFailed() is called
+// for the peer and it is no longer connected.
+//
+// Returns an empty ID if no appropriate peers are available.
+func (m *peerManager) DialNext() (NodeID, PeerAddress, error) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	peers, err := m.store.List()
+	if err != nil {
+		return "", PeerAddress{}, err
+	}
+	for _, peer := range peers {
+		switch {
+		case len(peer.Addresses) == 0:
+		case m.dialing[peer.ID]:
+		case m.connected[peer.ID]:
+		default:
+			// FIXME: We currently only dial the first address, but we should
+			// track connection statistics for each address and return the most
+			// appropriate one.
+			m.dialing[peer.ID] = true
+			return peer.ID, peer.Addresses[0], nil
+		}
+	}
+	return "", PeerAddress{}, nil
+}
+
+// DialFailed reports a failed dial attempt. This will make the peer available
+// for dialing again when appropriate.
+func (m *peerManager) DialFailed(peerID NodeID, address PeerAddress) error {
+	m.mtx.Lock()
```

Comment:
`address` is not used. Is it an argument to handle future stats?

Reply:
Yes, will be my next PR.